Sources/SwiftParserDiagnostics/ParseDiagnosticsGenerator.swift

```diff
@@ -171,6 +171,46 @@ public class ParseDiagnosticsGenerator: SyntaxAnyVisitor {
     return handledNodes.contains(node.id)
   }
 
+  /// A generalized version of `exchangeTokens` to support diagnostics with arbitrary syntax granularities not limited
+  /// to `TokenSyntax`.
+  ///
+  /// All nodes in `unexpected` that satisfy `unexpectedNodeCondition` will have  their present tokens flattened
+  /// into `misplacedTokens`.  All tokens in `correctNodes` will be flattened into `correctTokens`.  After that the
+  /// remaining operations will be the same as `exchangeTokens`.
+  public func exchangeNodes(
+    unexpected: UnexpectedNodesSyntax?,
+    unexpectedNodeCondition: (Syntax) -> Bool,
+    correctNodes: [Syntax?],
+    message: (_ misplacedTokens: [TokenSyntax]) -> some DiagnosticMessage,
+    moveFixIt: (_ misplacedTokens: [TokenSyntax]) -> FixItMessage,
+    removeRedundantFixIt: (_ misplacedTokens: [TokenSyntax]) -> FixItMessage? = { _ in nil }
+  ) {
```

Comment:
Instead of implementing `exchangeNodes` in terms of `exchangeTokens`, would it be possible to just generalize `exchangeTokens` to work `Syntax` instead of `TokenSyntax`? I briefly looked through it and I didn’t find any huge blockers but I might have missed something.